contributors/design-proposals/encryption.md

```diff
@@ -0,0 +1,204 @@
+
+The provider will be assigned a versioned identifier to uniquely pair the implementation with the data at rest, such as “k8s-aes-gcm-v1”.  Any implementation that attempts to decode data associated with this provider id must follow a known structure and apply a specific algorithm.
+
+The provider would take a set of keys and unique key identifiers from the command line, with the key values stored on disk.  One key is identified as the write key, all others are used to decrypt data from previous keys.  Keys must be rotated more often than every 2^32 writes
+
+The provider should use the recommended Go defaults for all crypto settings unless otherwise noted.  We should use AES-256 keys (32 bytes).
+
+Process for transforming a value (object encoded as JSON or protobuf) to and from stable storage will look like the following:
+
+Layout as written to etcd2 (json safe string only):
+```
+NONCE := read(/dev/urandom)
+PLAIN_TEXT := <VALUE>
+AUTHENTICATED_DATA := ETCD_KEY
+CIPHER_TEXT := aes_gcm_encrypt(KEY, IV:NONCE, PLAIN_TEXT, A:AUTHENTICATED_DATA)
+BASE64_DATA := base64(<NONCE><CIPHER_TEXT>)
+STORED_DATA := <PROVIDER>:<KEY_ID>:<BASE64_DATA>
+```
+
+Layout as written to etcd3 (bytes):
+```
+NONCE := read(/dev/urandom)
+PLAIN_TEXT := <VALUE>
+AUTHENTICATED_DATA := ETCD_KEY
+CIPHER_TEXT := aes_gcm_encrypt(KEY, IV:NONCE, PLAIN_TEXT, A:AUTHENTICATED_DATA)
+STORED_DATA := <PROVIDER_ID>:<KEY_ID>:<NONCE><AUTHENTICATED_DATA><CIPHER_TEXT>
+```
+
+Pseudo-code for encrypt (golang):
+```go
+block := aes.NewCipher(primaryKeyString)
+aead := cipher.NewGCM(c.block)
+keyId := primaryKeyId
+
+// for each write
+nonce := make([]byte, block.BlockSize())
+io.ReadFull(crypto_rand.Reader, nonce)
+authenticatedData := ETCD_KEY
+cipherText := aead.Seal(nil, nonce, value, authenticatedData)
+storedData := providerId + keyId + base64.Encode(nonce + authenticatedData + cipherText)
+```
+
+Pseudo-code for decrypt (golang):
+```go
+// for each read
+providerId, keyId, base64Encoded := // slice provider and key from value
+
+// ensure this provider is the one handling providerId
+aead := // lookup an aead instance for keyId or error
+bytes := base64.Decode(base64Encoded)
+nonce, authenticatedData, cipherText := // slice from bytes
+out, err := aead.Open(nil, nonce, cipherText, authenticatedData)
+```
+
+## Key Generation & Rotation
+
+To account for easy key rotation, an additional layer of abstraction is introduced where keys used for resource encryption is encrypted by another set of keys (Key Encryption Keys aka KEK).
+The key used for encrypting kubernetes resources (Data Encryption Keys, aka DEK) are generated by the apiserver and stored in etcd once encrypted with one of the KEKs.
+KEKs should never be stored in etcd and in most cases be on the same volume as the apiserver. This provides consistency for the apiserver but still protects against compromises to the etcd volume.
```

Comment:
I hope this section would apply for the "default" provider only because ...

Try to imagine the implementation of an HSM provider:
1. When you try to integrate with an HSM (e.g. with PKCS#11), the preferred way of encryption/decryption would be that you generate the DEK on the HSM, and also never export the DEK from the HSM.
2. Key rotation would mean, you generate a new key on the HSM and tell your API that you now have this new key
3. You would *not want* to let the DEKs being generated by the apiserver
4. There is only a need for a KEK if the DEKs are indeed generated by the apiserver and you basically only store the KEKs in your  HSM

Alternatively to support this approach an HSM provider implementation could look like:
- KEKs are generated and stored on the HSM
- kubelet startup loads them (extracts them - which means that it would need to get "wrapped" with the HSM (another KEK layer) and "unwrapped" in software again
- now this could decrypt the DEKs and basically use the default provider approach (golang AES-GCM) for doing the actual encryption/decryption work